File: src/main/java/cc/mirukuneko/realtrainmodrenewed/rail/math/StraightLine.java
package cc.mirukuneko.realtrainmodrenewed.rail.math;

/**
 * Port of jp.legacy.legacylib.math.StraightLine (XZ plane: first coord = Z, second = X).
 */
public final class StraightLine implements ILine {
    public final double startX;
    public final double startY;
    public final double endX;
    public final double endY;
    private final double length;
    private final double slopeAngle;
    private final double slope;
    private final double intercept;

    public StraightLine(double p1, double p2, double p3, double p4) {
        this.startX = p1;
        this.startY = p2;
        this.endX = p3;
        this.endY = p4;
        double dx = p3 - p1;
        double dy = p4 - p2;
        if (dx == 0.0D) {
            this.slope = Double.NaN;
            this.intercept = p1;
        } else {
            this.slope = dy / dx;
            this.intercept = p2 - this.slope * p1;
        }
        this.length = Math.sqrt(dx * dx + dy * dy);
        this.slopeAngle = Math.atan2(dy, dx);
    }

    @Override
    public double[] getPoint(int par1, int par2) {
        int i0 = par2 < 0 ? 0 : (par2 > par1 ? par1 : par2);
        double d0 = (double) i0 / (double) par1;
        double x = this.startX + (this.endX - this.startX) * d0;
        double y = this.startY + (this.endY - this.startY) * d0;
        return new double[]{x, y};
    }

    @Override
    public int getNearlestPoint(int par1, double y, double x) {
        double t;
        if (Double.isNaN(this.slope)) {
            t = (y - this.startY) / (this.endY - this.startY);
        } else {
            double a21 = 1.0D / (this.slope * this.slope + 1.0D);
            double x0 = (x + this.slope * y - this.slope * this.intercept) * a21;
            t = (x0 - this.startX) / (this.endX - this.startX);
        }
        return CurveMath.floor(t * (double) par1);
    }

    @Override
    public double getSlope(int par1, int par2) {
        return this.slopeAngle;
    }

    @Override
    public double getLength() {
        return this.length;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/rail/math/StraightLine.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.rail.math

import kotlin.math.*

/**
 * Port of jp.legacy.legacylib.math.StraightLine (XZ plane: first coord = Z, second = X).
 */
class StraightLine(
    @JvmField val startX: Double,
    @JvmField val startY: Double,
    @JvmField val endX: Double,
    @JvmField val endY: Double
) : ILine {
    private val slope: Double
    private val intercept: Double
    override val length: Double
    private val slopeAngle: Double

    init {
        val dx = endX - startX
        val dy = endY - startY
        if (dx == 0.0) {
            slope = Double.NaN
            intercept = startX
        } else {
            slope = dy / dx
            intercept = startY - slope * startX
        }
        length = sqrt(dx * dx + dy * dy)
        slopeAngle = atan2(dy, dx)
    }

    override fun getPoint(split: Int, index: Int): DoubleArray {
        val i0 = if (index < 0) 0 else (if (index > split) split else index)
        val d0 = i0.toDouble() / split.toDouble()
        val x = startX + (endX - startX) * d0
        val y = startY + (endY - startY) * d0
        return doubleArrayOf(x, y)
    }

    override fun getNearlestPoint(split: Int, x: Double, z: Double): Int {
        val t: Double = if (slope.isNaN()) {
            (x - startY) / (endY - startY)
        } else {
            val a21 = 1.0 / (slope * slope + 1.0)
            val x0 = (z + slope * x - slope * intercept) * a21
            (x0 - startX) / (endX - startX)
        }
        return CurveMath.floor(t * split.toDouble())
    }

    override fun getSlope(split: Int, index: Int): Double = slopeAngle
}
